### PR TITLE
add memberOf to /users endpoint and members to /groups endpoint

### PR DESCRIPTION
**Pull request by dragonchaser**
This PR adds a members slice to the `/groups` endpoint and
a members slice to the `/users` endpoint.

refs #3837

### CODE REVIEW

**Comment by ownclouders** [comment]
:boom: Acceptance test [<strong>Core-API-Tests-ocis-storage-4</strong>](https://drone.owncloud.com/owncloud/ocis/12351/42/1) failed. Further test are cancelled...

File: extensions/graph/pkg/identity/ldap.go [hunk 1]
```diff
@@ -5,12 +5,15 @@ import (
...
 	"net/url"
+	"strings"
 
 	"github.com/go-ldap/ldap/v3"
 	"github.com/gofrs/uuid"
 	ldapdn "github.com/libregraph/idm/pkg/ldapdn"
 	libregraph "github.com/owncloud/libre-graph-api-go"
 
+	"golang.org/x/exp/slices"
+
```
Comment (rhafer) [inline extensions/graph/pkg/identity/ldap.go:16]: I don't think we need the empty lines here. I guess the one on line 14 got also there accidently. (At least elsewhere we only split the `import` block into two parts)

File: extensions/graph/pkg/identity/ldap.go [hunk 2]
```diff
@@ -352,13 +355,32 @@ func (i *LDAP) getLDAPUserByFilter(filter string) (*ldap.Entry, error) {
...
 
-func (i *LDAP) GetUser(ctx context.Context, nameOrID string) (*libregraph.User, error) {
+func (i *LDAP) GetUser(ctx context.Context, nameOrID string, queryParam url.Values) (*libregraph.User, error) {
 	i.logger.Debug().Str("backend", "ldap").Msg("GetUser")
 	e, err := i.getLDAPUserByNameOrID(nameOrID)
 	if err != nil {
 		return nil, err
 	}
-	return i.createUserModelFromLDAP(e), nil
+	userGroups, err := i.getGroupsForUser(e.DN)
```
Comment (rhafer) [inline extensions/graph/pkg/identity/ldap.go:364]: This query should go into the `if slices.Contains(sel, "memberOf") || slices.Contains(exp, "memberOf")` block. We should only do it if group memberships are explicitly requested.

**Review by rhafer — CHANGES_REQUESTED** [review]
Just a few minor things. Looks good to me otherwise.

**Comment by phil-davis** [comment]
@dragonchaser I have built an ocis with the latest master and run it:
```
make clean
make generate
cd ocis
make build
cd bin
PROXY_ENABLE_BASIC_AUTH=true OCIS_INSECURE=true ./ocis server
```

But my attempts to use the new member API queries are not successful:
```
$ curl -k 'https://localhost:9200/graph/v1.0/users' -u admin:admin
{"value":[{"displayName":"Admin","id":"473ecfbe-1137-4793-8a8e-14cc2d1e5b1e","mail":"admin@example.org","onPremisesSamAccountName":"admin"},{"displayName":"Phil Davis","id":"bad24955-8c2a-4bf6-a13e-d14ec2a587d1","mail":"phil@example.com","onPremisesSamAccountName":"phil"},{"displayName":"Mr Test","id":"654f7536-c1f5-4e06-bbc5-d1630e452956","mail":"test@example.com","onPremisesSamAccountName":"test123"}]}
$ curl -k 'https://localhost:9200/graph/v1.0/groups' -u admin:admin
{"value":[{"displayName":"g42","id":"31572dac-a4b5-419b-b2e5-3390e6144386"},{"displayName":"z999","id":"26d47151-ecb2-4d82-a82b-01444998068d"}]}
$ curl -k 'https://localhost:9200/graph/v1.0/groups/31572dac-a4b5-419b-b2e5-3390e6144386/members' -u admin:admin
[{"displayName":"Phil Davis","id":"bad24955-8c2a-4bf6-a13e-d14ec2a587d1","mail":"phil@example.com","onPremisesSamAccountName":"phil"}]
```

`phil` ('bad24955-8c2a-4bf6-a13e-d14ec2a587d1`) is in group `g42` (`31572dac-a4b5-419b-b2e5-3390e6144386`)

But when I try to get the group(s) along with the members:
```
$ curl -k 'https://localhost:9200/graph/v1.0/groups?$expand=members' -u admin:admin
{"value":[{"displayName":"g42","id":"31572dac-a4b5-419b-b2e5-3390e6144386"},{"displayName":"z999","id":"26d47151-ecb2-4d82-a82b-01444998068d"}]}
```
I just get a response with the group display names.

Or query a single group:
```
$ curl -k 'https://localhost:9200/graph/v1.0/groups/31572dac-a4b5-419b-b2e5-3390e6144386/?$expand=members' -u admin:admin
{"displayName":"g42","id":"31572dac-a4b5-419b-b2e5-3390e6144386"}
```
I just get a response with the group display name.

Am I missing something?

**Comment by dragonchaser** [comment]
@phil-davis could not reproduce

```bash
$> curl -k $ curl -k 'https://localhost:9200/graph/v1.0/groups/509a9dcd-bb37-4f4f-a01a-19dca27d9cfa?$expand=members' -u admin:admin|jq
{
  "displayName": "users",
  "id": "509a9dcd-bb37-4f4f-a01a-19dca27d9cfa",
  "members": [
    {
      "displayName": "Albert Einstein",
      "id": "4c510ada-c86b-4815-8820-42cdf82c3d51",
      "mail": "einstein@example.org",
      "onPremisesSamAccountName": "einstein"
    },
    {
      "displayName": "Marie Skłodowska Curie",
      "id": "f7fbf8c8-139b-4376-b307-cf0a8c2d0d9c",
      "mail": "marie@example.org",
      "onPremisesSamAccountName": "marie"
    },
    {
      "displayName": "Richard Phillips Feynman",
      "id": "932b4540-8d16-481e-8ef4-588e4b6b151c",
      "mail": "richard@example.org",
      "onPremisesSamAccountName": "richard"
    },
    {
      "displayName": "Maurice Moss",
      "id": "058bff95-6708-4fe5-91e4-9ea3d377588b",
      "mail": "moss@example.org",
      "onPremisesSamAccountName": "moss"
    },
    {
      "displayName": "Admin",
      "id": "some-admin-user-id-0000-000000000000",
      "mail": "admin@example.org",
      "onPremisesSamAccountName": "admin"
    }
  ]
}
```

Can you run `ocis version` and paste the output?

**Comment by phil-davis** [comment]
@dragonchaser I pulled current master and rebuilt my oCIS:
```
$ ./ocis version
Version: 2.0.0-beta.3+ab221d2b2
Compiled: 2022-06-14 00:00:00 +0000 UTC

No running services found.
```

And I still have the same problem.

**Comment by phil-davis** [comment]
I added another user to group "g42". Now all the `expand=members` and `expand=memberOf` queries are working. 
```
curl -k --header "Content-Type: application/json" \
        --request POST  --data \
	'{ "@odata.id": "https://localhost:9200/graph/v1.0/users/acc15432-9dc4-424c-91d7-6867572a244c" }' \
	'https://localhost:9200/graph/v1.0/groups/31572dac-a4b5-419b-b2e5-3390e6144386/members/$ref' -u admin:admin
```

Probably I already had user `phil` in group `g42` before this new feature was added to oCIS. "something" seems to have "rebuilt" the user<=>group relationship information "somewhere" and it works now.

**Comment by dragonchaser** [comment]
Confusing, there have been no changes to these structures, the change just adds LDAP queries to the code.

**Pull request merged**